### PR TITLE
add interface for user defined GMG constraints

### PR DESCRIPTION
**Pull request by tcclevenger**
Add an interface for the user to add in their own constraints for any given level. Only implemented to work with `MGTransferMatrixFree` class. Implementation: before prolongation, set the correct constrained values.

We use this inside new code going into ASPECT, namely creating no-normal-flux constraints for a sphere. And it has been verified to work inside the GMG preconditioner for those problems (constant/low iteration counts with refinement, parallel).

### CODE REVIEW

File: include/deal.II/multigrid/mg_constrained_dofs.h [hunk 1]
```diff
@@ -103,6 +103,15 @@ class MGConstrainedDoFs : public Subscriptor
     const std::set<types::boundary_id> &boundary_ids,
     const ComponentMask &               component_mask = ComponentMask());
 
+  /**
+   * Set user defined contraints to be used only in the MGTransferMatrixFree
+   * class. The user can add multiple constraints, but they will be
+   * merged such that the current merge will win any conflicts.
```
Comment (tjhei) [inline include/deal.II/multigrid/mg_constrained_dofs.h]: suggestion:
```
   * Set user defined constraints to be used on level @p level.
   *
   * The user can add ...
   * @note This is currently only implemented for MGTransferMatrixFree.
```

What do you mean by "can add"? Do you mean "The function can be called more than once per level and constraints will be merged with the existing constraints"? If yes, how do you reset the constraints? Why is the function not called add_user_constraints?

Please also document how this interacts with Dirichlet constraints.

File: include/deal.II/multigrid/mg_constrained_dofs.h [hunk 2]
```diff
@@ -196,6 +205,12 @@ class MGConstrainedDoFs : public Subscriptor
   const AffineConstraints<double> &
   get_level_constraint_matrix(const unsigned int level) const;
 
+  /**
+   * Return the user defined constraint matrix for a given level.
+   */
```
Comment (tjhei) [inline include/deal.II/multigrid/mg_constrained_dofs.h]: Mention that it is filled by set_user_constraints() and that is does not contain dirichlet values. or does it?

File: source/lac/affine_constraints.inst.in [hunk 3]
```diff
@@ -344,4 +344,8 @@ for (T : DEAL_II_VEC_TEMPLATES)
       T<float> &) const;
     template void dealii::AffineConstraints<double>::distribute<T<float>>(
       T<float> &) const;
+
+    template void dealii::AffineConstraints<double>::distribute<
```
Comment (tjhei) [inline source/lac/affine_constraints.inst.in]: unrelated change?

File: source/multigrid/mg_transfer_matrix_free.cc [hunk 4]
```diff
@@ -355,6 +355,24 @@ MGTransferMatrixFree<dim, Number>::do_prolongate_add(
...
+  LinearAlgebra::distributed::Vector<Number> new_src;
+  {
+    LinearAlgebra::distributed::Vector<Number> copy_src(src);
+
+    // Distribute any user defined constraints
+    this->mg_constrained_dofs->get_user_constraint_matrix(to_level - 1)
+      .distribute(copy_src);
+
+    // Create a new ghosted vector from the copied source vector
+    new_src.reinit(copy_src);
```
Comment (tjhei) [inline source/multigrid/mg_transfer_matrix_free.cc]: Can we skip doing this work if no user constraints have been set? This might be difficult to detect as only some of the processors have entries. Maybe only do the .reinit for the constraintmatrix if the user adds constraints? Then you can check the .size() here.

File: include/deal.II/multigrid/mg_constrained_dofs.h [hunk 5]
```diff
@@ -487,6 +542,15 @@ MGConstrainedDoFs::get_level_constraint_matrix(const unsigned int level) const
 
 
 
+inline const AffineConstraints<double> &
+MGConstrainedDoFs::get_user_constraint_matrix(const unsigned int level) const
+{
+  AssertIndexRange(level, user_constraints.size());
+  return user_constraints[level];
```
Comment (tjhei) [inline include/deal.II/multigrid/mg_constrained_dofs.h]: also assert that ``user_constraints[level].size()>0``

File: source/multigrid/mg_transfer_matrix_free.cc (hunk 4 above)
Comment (tjhei) [reply source/multigrid/mg_transfer_matrix_free.cc]: Not quite. ``.get_local_lines().size() > 0`` checks if this processor has any constraints and not if custom user constraints are in use. You can check this with ``.size()>0`` with your change to add_user_constraints.
Finally, we still have the problem that you made the code path without constraints slower.
I would do something like:
```
Vector new_src;
Vector* to_use = &src;
if (user_constraints[level].size()>0)
{
// initialize new_src
to_use = &new_src;
}

// use *to_use down here
```

File: include/deal.II/multigrid/mg_constrained_dofs.h (hunk 5 above)
Comment (tcclevenger) [reply include/deal.II/multigrid/mg_constrained_dofs.h]: I think we don't want to assert here since we use an empty constraint matrix in the transfer to decide not to reinitialize vectors.

Comment (tjhei) [reply include/deal.II/multigrid/mg_constrained_dofs.h]: oh, sure. That makes sense.

**Comment by tcclevenger** [comment]
Some of the tests were failing because `this->mg_constrained_dofs` is a `nullptr` if its not set in the construction of the transfer object. I added the check `this->mg_constrained_dofs != nullptr` in the if-statement in mg_transfer_matrix_free.cc.

File: source/multigrid/mg_transfer_matrix_free.cc [hunk 6]
```diff
@@ -355,6 +355,30 @@ MGTransferMatrixFree<dim, Number>::do_prolongate_add(
     Utilities::fixed_power<dim>(n_child_dofs_1d);
   constexpr unsigned int three_to_dim = Utilities::pow(3, dim);
 
+  // If we have user defined MG constraints, we must create
+  // a non-const, ghosted version of the source vector to distirbute
+  // constraints.
+  const LinearAlgebra::distributed::Vector<Number> *to_use = &src;
+  LinearAlgebra::distributed::Vector<Number>        new_src;
+  if (this->mg_constrained_dofs != nullptr &&
```
Comment (tjhei) [inline source/multigrid/mg_transfer_matrix_free.cc]:
```suggestion
  if ((*this->mg_constrained_dofs) != nullptr &&
```

Comment (tcclevenger) [reply source/multigrid/mg_transfer_matrix_free.cc]: We get the following compile error with this change:

error1

Comment (peterrum) [reply source/multigrid/mg_transfer_matrix_free.cc]: @tjhei why do you want to dereference the pointer?

**Comment by tcclevenger** [comment]
The tests aren't passing because when constructing the transfer, they did not initialize an `MGConstrainedDoFs` object. I thought adding `if (this->mg_constrained_dofs != nullptr)` to the prolongation (line 363 in mg_transfer_matrix_free.cc) would fix this, but the tests fail with the following error:
error

Any suggestions?

**Comment by peterrum** [comment]
try out these changes:
```diff
diff --git a/source/multigrid/mg_transfer_matrix_free.cc b/source/multigrid/mg_transfer_matrix_free.cc
index b1876a7..feecf04 100644
--- a/source/multigrid/mg_transfer_matrix_free.cc
+++ b/source/multigrid/mg_transfer_matrix_free.cc
@@ -104,7 +104,7 @@ MGTransferMatrixFree<dim, Number>::build(const DoFHandler<dim, dim> &mg_dof)
 
   internal::MGTransfer::setup_transfer<dim, Number>(
     mg_dof,
-    this->mg_constrained_dofs,
+    this->mg_constrained_dofs == nullptr ? nullptr : this->mg_constrained_dofs,
     elem_info,
     level_dof_indices,
     parent_child_connect,
diff --git a/source/multigrid/mg_transfer_prebuilt.cc b/source/multigrid/mg_transfer_prebuilt.cc
index 74c60e7..9085ae5 100644
--- a/source/multigrid/mg_transfer_prebuilt.cc
+++ b/source/multigrid/mg_transfer_prebuilt.cc
@@ -105,10 +105,13 @@ MGTransferPrebuilt<VectorType>::prolongate(const unsigned int to_level,
 {
   Assert((to_level >= 1) && (to_level <= prolongation_matrices.size()),
          ExcIndexRange(to_level, 1, prolongation_matrices.size() + 1));
+#ifdef DEBUG
+  if(this->mg_constrained_dofs != nullptr)
   Assert(this->mg_constrained_dofs->get_user_constraint_matrix(to_level - 1)
              .get_local_lines()
              .size() == 0,
          ExcNotImplemented());
+#endif
 
   prolongation_matrices[to_level - 1]->vmult(dst, src);
 }
diff --git a/source/multigrid/multigrid.cc b/source/multigrid/multigrid.cc
index 1fbba5f..8607ab5 100644
--- a/source/multigrid/multigrid.cc
+++ b/source/multigrid/multigrid.cc
@@ -93,10 +93,13 @@ MGTransferBlock<number>::prolongate(const unsigned int         to_level,
          ExcDimensionMismatch(src.n_blocks(), this->n_mg_blocks));
   Assert(dst.n_blocks() == this->n_mg_blocks,
          ExcDimensionMismatch(dst.n_blocks(), this->n_mg_blocks));
+#ifdef DEBUG
+  if(this->mg_constrained_dofs != nullptr)
   Assert(this->mg_constrained_dofs->get_user_constraint_matrix(to_level - 1)
              .get_local_lines()
              .size() == 0,
          ExcNotImplemented());
+#endif
 
   // Multiplicate with prolongation
   // matrix, but only those blocks
@@ -283,10 +286,13 @@ MGTransferBlockSelect<number>::prolongate(const unsigned int    to_level,
 {
   Assert((to_level >= 1) && (to_level <= prolongation_matrices.size()),
          ExcIndexRange(to_level, 1, prolongation_matrices.size() + 1));
+  #ifdef DEBUG
+  if(this->mg_constrained_dofs != nullptr)
   Assert(this->mg_constrained_dofs->get_user_constraint_matrix(to_level - 1)
              .get_local_lines()
              .size() == 0,
          ExcNotImplemented());
+#endif
 
   prolongation_matrices[to_level - 1]
     ->block(selected_block, selected_block)
```

**Comment by tcclevenger** [comment]
Thanks @peterrum ! I forgot I was using the function elsewhere.

**Comment by bangerth** [comment]
Wait, @peterrum, isn't your proposed change
```
  this->mg_constrained_dofs == nullptr ? nullptr : this->mg_constrained_dofs
```
simply the same as saying
```
  this->mg_constrained_dofs
```
? :-)

File: source/multigrid/mg_transfer_matrix_free.cc [hunk 7]
```diff
@@ -355,6 +355,30 @@ MGTransferMatrixFree<dim, Number>::do_prolongate_add(
     Utilities::fixed_power<dim>(n_child_dofs_1d);
   constexpr unsigned int three_to_dim = Utilities::pow(3, dim);
 
+  // If we have user defined MG constraints, we must create
+  // a non-const, ghosted version of the source vector to distirbute
```
Comment (tjhei) [inline source/multigrid/mg_transfer_matrix_free.cc]:
```suggestion
  // a non-const, ghosted version of the source vector to distribute
```

**Comment by peterrum** [comment]
> Wait, @peterrum, isn't your proposed change
> 
> ```
>   this->mg_constrained_dofs == nullptr ? nullptr : this->mg_constrained_dofs
> ```
> 
> simply the same as saying
> 
> ```
>   this->mg_constrained_dofs
> ```
> 
> ? :-)

@bangerth Apparently it is the same ;) I would have been surprised it would be otherwise (what the class does is clear). I started debugging with that line because of @tjhei and @tcclevenger's discussion and continued from there... I have rerun the tests on our cluster and it works without that change!
